internal/controllers/gateway/gateway_controller.go

```diff
@@ -105,6 +105,13 @@ func (r *GatewayReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		return err
 	}
 
+	if err := c.Watch(
+		&source.Kind{Type: &gatewayv1alpha2.ReferenceGrant{}},
+		handler.EnqueueRequestsFromMapFunc(r.listReferenceGrantsForGateway),
```

Comment:
since we do not care about `ReferenceGrants` has no `Gateway` in its `From`, maybe here we can add a predicate function to reduce number of ineffective watched events.